cpp/src/arrow/util/functional.h

```diff
@@ -79,5 +84,47 @@ struct call_traits {
       typename std::enable_if<std::is_same<return_type<F>, T>::value, RT>;
 };
 
+/// A type erased callable object which may only be invoked once.
```

Comment:
Is it a kind of lighter-weight `std::function`?

Reply:
yes, with the added guarantee that when it is called it will drop all references which it has in closure